### PR TITLE
Smart open to transfer datasets for universal transfer operator 

### PR DESCRIPTION
**Pull request by sunank200**
# Description
## What is the current behavior?

Currently, all the files are being downloaded to temporary reference in the worker and uploaded back to the destination dataset. This is time-consuming and doesn't work well with database transfers. Hence, implemented the transfer for all file types using smart open.

closes: #1619 

## What is the new behavior?

- Add a feature to read the file using smart open and convert it into a buffer that can be streamed into other data structures.
- Run the example DAG for S3 and GCS transfers.
- Add class for `Filetype` and `PandasDataframe`.
- For filetypes CSV, json and ndjson, read as `io.StringIO()`. For other filetypes read as `io.BytesIO()`

## Does this introduce a breaking change?
No

### Checklist
- [x] Ran example DAG successfully.
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: universal_transfer_operator/src/universal_transfer_operator/datasets/dataframe/pandas.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    :param df: dataframe whose cols will be altered
+    :param columns_names_capitalization: String Literal with possible values - lower/Upper
+    """
+    if isinstance(df, pd.DataFrame):
+        df = PandasDataframe.from_pandas_df(df)
+        if columns_names_capitalization == "lower":
+            df.columns = [col_label.lower() for col_label in df.columns]  # skipcq: PYL-W0201
+        elif columns_names_capitalization == "upper":
+            df.columns = [col_label.upper() for col_label in df.columns]  # skipcq: PYL-W0201
```
Comment (dimberman) [inline universal_transfer_operator/src/universal_transfer_operator/datasets/dataframe/pandas.py:100]: Is there no more option for original?

File: universal_transfer_operator/src/universal_transfer_operator/datasets/dataframe/pandas.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+):
+    """
+    Convert cols of a dataframe to required case. Options - lower/Upper
+
+    :param df: dataframe whose cols will be altered
+    :param columns_names_capitalization: String Literal with possible values - lower/Upper
+    """
+    if isinstance(df, pd.DataFrame):
+        df = PandasDataframe.from_pandas_df(df)
+        if columns_names_capitalization == "lower":
```
Comment (dimberman) [inline universal_transfer_operator/src/universal_transfer_operator/datasets/dataframe/pandas.py:97]: Should this not be a Literal?

**Review by dimberman — APPROVED** [review]
Overall looks good, just some minor comments

File: universal_transfer_operator/src/universal_transfer_operator/datasets/dataframe/pandas.py (hunk 1 above)
Comment (sunank200) [reply universal_transfer_operator/src/universal_transfer_operator/datasets/dataframe/pandas.py:100]: This is what we have in python-sdk as well

File: universal_transfer_operator/src/universal_transfer_operator/datasets/dataframe/pandas.py (hunk 2 above)
Comment (sunank200) [reply universal_transfer_operator/src/universal_transfer_operator/datasets/dataframe/pandas.py:97]: This is what we have in python-sdk as well.

File: universal_transfer_operator/src/universal_transfer_operator/datasets/file/types/ndjson.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+from universal_transfer_operator.constants import DEFAULT_CHUNK_SIZE, FileType as FileTypeConstants
+from universal_transfer_operator.datasets.dataframe.pandas import (
+    PandasDataframe,
+    convert_columns_names_capitalization,
+)
+from universal_transfer_operator.datasets.file.types.base import FileType
+
+
+class NDJSONFileType(FileType):
+    """Concrete implementation to handle NDJSON file type"""
```
[the remark this answers is not in the document]
Comment (sunank200) [reply universal_transfer_operator/src/universal_transfer_operator/datasets/file/types/ndjson.py]: https://github.com/astronomer/astro-sdk/pull/1621/commits/8a78cc81f2fcf4cca47f7826eed8c478a60cd753

File: universal_transfer_operator/src/universal_transfer_operator/datasets/file/types/ndjson.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+from universal_transfer_operator.constants import DEFAULT_CHUNK_SIZE, FileType as FileTypeConstants
+from universal_transfer_operator.datasets.dataframe.pandas import (
+    PandasDataframe,
+    convert_columns_names_capitalization,
+)
+from universal_transfer_operator.datasets.file.types.base import FileType
+
+
+class NDJSONFileType(FileType):
```
[the remark this answers is not in the document]
Comment (sunank200) [reply universal_transfer_operator/src/universal_transfer_operator/datasets/file/types/ndjson.py]: https://github.com/astronomer/astro-sdk/pull/1621/commits/8a78cc81f2fcf4cca47f7826eed8c478a60cd753

**Pull request merged**